File: Baekjoon/suah/포도주 시식.py
import sys
input = lambda : sys.stdin.readline().strip()

n = int(input())
wine = []
for _ in range(n):
    wine.append(int(input()))

dp = [0] * n
dp[0] = wine[0]
if n > 1:
    dp[1] = wine[0] + wine[1]
if n > 2:
    dp[2] = max(wine[2]+wine[0], wine[2] + wine[1], dp[1])
for i in range(3, n):
    dp[i] = max(dp[i-3] + wine[i-1] + wine[i], dp[i-2] + wine[i], dp[i-1])

print(dp[n-1])

File: Baekjoon/yerim/포도주 시식.cpp
// [2156] 포도주 시식
#include <iostream>
using namespace std;

int n, dp[10001], wine[10001];

int main(void) {
    cin >> n;
    for(int i = 1; i <= n; i++) {
        cin >> wine[i];
    }

    dp[1] = wine[1];
    dp[2] = wine[1] + wine[2];

    for (int i = 3; i <= n; i++) {
        dp[i] = wine[i] + max(dp[i - 2], wine[i - 1] + dp[i - 3]);
        dp[i] = max(dp[i], dp[i - 1]);
    }

    cout << dp[n] << '\n';
    return 0;
}